tests/dragonfly/replication_test.py

```diff
@@ -410,50 +410,43 @@ async def test_rotating_masters(
 @pytest.mark.slow
 async def test_cancel_replication_immediately(df_local_factory, df_seeder_factory):
     """
-    Issue 80 replication commands randomally distributed over 10 seconds. This
-    checks that the replication state machine can handle cancellation well.
-    We assert that at least one command was cancelled during start and at least
-    one command was successfull.
+    Issue 100 replication commands. This checks that the replication state
+    machine can handle cancellation well.
+    We assert that at least one command was cancelled.
     After we finish the 'fuzzing' part, replicate the first master and check that
     all the data is correct.
     """
-    COMMANDS_TO_ISSUE = 80
+    COMMANDS_TO_ISSUE = 100
 
     replica = df_local_factory.create()
-    masters = [df_local_factory.create() for i in range(4)]
-    df_local_factory.start_all([replica] + masters)
+    master = df_local_factory.create()
+    df_local_factory.start_all([replica, master])
 
-    seeders = [df_seeder_factory.create(port=m.port) for m in masters]
+    seeder = df_seeder_factory.create(port=master.port)
     c_replica = aioredis.Redis(port=replica.port)
-    await asyncio.gather(*(seeder.run(target_deviation=0.1) for seeder in seeders))
+    await seeder.run(target_deviation=0.1)
 
     replication_commands = []
 
-    async def replicate(index):
-        await asyncio.sleep(10.0 * random.random())
```

Comment:
because you removed the sleep, now we tests a very specific scenario I believe that we execute replicaof on replica and its imedietly gets cancelled, we are not getting to the greeting not sure even if dns resolve is done.

Reply:
This test was never about reaching Greeting, it was about checking if the context got properly cancelled: 

See start:

```
    79 error_code Replica::Start(ConnectionContext* cntx) {                                              
    80   VLOG(1) << "Starting replication";                                                              
    81   ProactorBase* mythread = ProactorBase::me();                                                    
    82   CHECK(mythread);                                                                                
    83                                                                                                   
    84   RETURN_ON_ERR(cntx_.SwitchErrorHandler(absl::bind_front(&Replica::DefaultErrorHandler, this))); 
    85                                                                                                   
    86   auto check_connection_error = [this, &cntx](const error_code& ec, const char* msg) -> error_code {
    87     if (cntx_.IsCancelled()) {                                                                    
    88       (*cntx)->SendError("replication cancelled");                                                
    89       return std::make_error_code(errc::operation_canceled);                                      
    90     } 
    91     if (ec) {                                                                                     
    92       (*cntx)->SendError(absl::StrCat(msg, ec.message()));                                        
    93       cntx_.Cancel();
    94       return ec;
    95     } 
    96     return {};                                                                                    
    97   };
    98   
    99   // 1. Resolve dns.                                                                              
   100   VLOG(1) << "Resolving master DNS";                                                              
   101   error_code ec = ResolveMasterDns();                                                             
   102   RETURN_ON_ERR(check_connection_error(ec, "could not resolve master dns"));                      
   103   
   104   // 2. Connect socket.                           
   105   VLOG(1) << "Connecting to master";                                                
   106   ec = ConnectAndAuth(absl::GetFlag(FLAGS_master_connect_timeout_ms) * 1ms, &cntx_);
   107   RETURN_ON_ERR(check_connection_error(ec, kConnErr));
   108   
   109   // 3. Greet.                                   
   110   VLOG(1) << "Greeting";                                            
   111   state_mask_.store(R_ENABLED | R_TCP_CONNECTED);                                                 
   112   ec = Greet();
   113   RETURN_ON_ERR(check_connection_error(ec, "could not greet master "));
   114 
   115   // 4. Spawn main coordination fiber.
   116   sync_fb_ = fb2::Fiber("main_replication", &Replica::MainReplicationFb, this);
```

We explicitly looked for `replication cancelled` in this test which can happen in any of the 4 steps above.

Reply:
Basically the next `ReplicaOf` command cancels the previous one see server_family.cc
```
  1679   if (replica_) {
  1680     replica_->Stop();  // NOTE: consider introducing update API flow.                             
  1681   } else {

```

Reply:
would calling "info replication" and checking the state help this?

Reply:
I am not quite sure how, `INFO REPLCATION` would return information about the status of the replica for example:
```
# Replication
role:replica
master_host:localhost
master_port:6379
master_link_status:up
master_last_io_seconds_ago:0
master_sync_in_progress:0
```

What we are interested in is to trigger a context cancellation and the way we do that is by issuing two `ReplicaOf` commands as fast as possible such that the a second call to `ReplicaOf` would cancel the previous one (and send the error message that this test searches for)

Reply:
how do send two command as fast as possible? do we use separate client connections to the same replica for this?